# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns; sns.set()

# Challenge 1 - Explore the Scikit-Learn Datasets

Before starting to work on our own datasets, let's first explore the datasets that are included in this Python library. These datasets have been cleaned and formatted for use in ML algorithms.

First, we will load the diabetes dataset. Do this in the cell below by importing the datasets and then loading the dataset  to the `diabetes` variable using the `load_diabetes()` function ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_diabetes.html)).

In [2]:
from sklearn.datasets import load_diabetes

diabetes = load_diabetes()

Let's explore this variable by looking at the different attributes (keys) of `diabetes`. Note that the `load_diabetes` function does not return dataframes. It returns you a Python dictionary.

In [3]:
diabetes.keys()

dict_keys(['data', 'target', 'DESCR', 'feature_names', 'data_filename', 'target_filename'])

#### The next step is to read the description of the dataset. 

Print the description in the cell below using the `DESCR` attribute of the `diabetes` variable. Read the data description carefully to fully understand what each column represents.

*Hint: If your output is ill-formatted by displaying linebreaks as `\n`, it means you are not using the `print` function.*

In [4]:
print(diabetes['DESCR'])

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - Age
      - Sex
      - Body mass index
      - Average blood pressure
      - S1
      - S2
      - S3
      - S4
      - S5
      - S6

Note: Each of these 10 feature variables have been mean centered and scaled by the standard deviation times `n_samples` (i.e. the sum of squares of each column totals 1).

Source URL:
https://www4.stat.ncsu.edu/~boos/var.select/diabetes.html

For more information see:
Bra

#### Based on the data description, answer the following questions:

1. How many attributes are there in the data? What do they mean?

1. What is the relation between `diabetes['data']` and `diabetes['target']`?

1. How many records are there in the data?

- 1.Tiene 10 atributos (10 columnas), son las variables o features del dataframe
- 2.Diabetes['target'] es nuestra "y", es decir, lo que queremos predecir o identificar, mientras que "X" son las variables de describen, explican o determinan a "y"
- 3.Contiene 442 records, instancias o 'filas'

#### Now explore what are contained in the *data* portion as well as the *target* portion of `diabetes`. 

Scikit-learn typically takes in 2D numpy arrays as input (though pandas dataframes are also accepted). Inspect the shape of `data` and `target`. Confirm they are consistent with the data description.

In [5]:
diabetes['data'].shape, diabetes['target'].shape

((442, 10), (442,))

In [6]:
#Concuerda con las descripcion de los datos

# Challenge 2 - Perform Supervised Learning on the Dataset

The data have already been split to predictor (*data*) and response (*target*) variables. Given this information, we'll apply what we have previously learned about linear regression and apply the algorithm to the diabetes dataset.

#### Let's briefly revisit the linear regression formula:

```
y = β0 + β1X1 + β2X2 + ... + βnXn + ϵ
```

...where:

- X1-Xn: data 
- β0: intercept 
- β1-βn: coefficients 
- ϵ: error (cannot explained by model)
- y: target

Also take a look at the `sklearn.linear_model.LinearRegression` [documentation](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html).

#### In the cell below, import the `linear_model` class from `sklearn`. 

In [7]:
from sklearn.linear_model import LinearRegression as LinReg

#### Create a new instance of the linear regression model and assign the new instance to the variable `diabetes_model`.

In [8]:
diabetes_model = LinReg()

#### Next, let's split the training and test data.

Define `diabetes_data_train`, `diabetes_target_train`, `diabetes_data_test`, and `diabetes_target_test`. Use the last 20 records for the test data and the rest for the training data.

In [9]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(diabetes['data'], diabetes['target'], test_size = 0.2)

Fit the training data and target to `diabetes_model`. Print the *intercept* and *coefficients* of the model.

In [10]:
diabetes_model.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [11]:
diabetes_model.intercept_

151.31015555772598

In [12]:
diabetes_model.coef_

array([ -43.50926294, -224.32655581,  526.55695437,  313.97319442,
       -543.9734345 ,  270.99412821,   59.10566019,  229.30670781,
        660.67696993,   80.50787139])

#### Inspecting the results

From the outputs you should have seen:

- The intercept is a float number.
- The coefficients are an array containing 10 float numbers.

This is the linear regression model fitted to your training dataset.

#### Using your fitted linear regression model, predict the *y* of `diabetes_data_test`.

In [13]:
diabetes_model.predict(X_test)

array([103.82545158, 190.64597486, 226.34700397,  92.96690431,
       121.60741718, 158.8423593 , 215.9211228 , 142.50508118,
        72.3828566 , 243.50081184,  58.12306482, 167.55821916,
       194.09823627, 240.0668976 , 246.40473586, 178.09805084,
       246.05650028,  80.019767  , 102.26203599, 179.21791017,
       115.33644534, 199.22094035, 140.66025791, 137.94498971,
       108.2597375 , 100.83613533, 101.98306858,  74.2711217 ,
       202.26160372, 135.30317284, 212.58126427, 141.57203454,
       142.76208368,  98.84088791,  78.58101255, 111.19097195,
       292.68026262,  45.14735263, 292.59650073, 146.11450198,
       159.43149794, 181.29132902,  79.21751762, 245.07488116,
       166.55941739, 186.60599923, 159.42293744, 122.42648484,
        49.74582324, 103.1241127 ,  58.23111873, 140.37889308,
       219.36710055, 117.95693646, 141.14460966,  78.49538135,
       195.8007518 , 155.58869616,  65.05041878, 161.54827384,
       154.85268158,  82.64769963, 260.41316643,  70.67

#### Print your `diabetes_target_test` and compare with the prediction. 

In [14]:
diab_preds = pd.DataFrame({'Real': y_test, 'Predicted': diabetes_model.predict(X_test)}).reset_index(drop=True)
diab_preds.head(15)

Real   Predicted
0    97.0  103.825452
1   283.0  190.645975
2   232.0  226.347004
3   101.0   92.966904
4    53.0  121.607417
5   259.0  158.842359
6   275.0  215.921123
7    88.0  142.505081
8    89.0   72.382857
9   264.0  243.500812
10   70.0   58.123065
11  128.0  167.558219
12  122.0  194.098236
13  217.0  240.066898
14  274.0  246.404736

In [15]:
print(f"Model performance on trainng set: {diabetes_model.score(X_train, y_train)}")
print(f"Model performance on testing set: {diabetes_model.score(X_test, y_test)}")

Model performance on trainng set: 0.49179454185888927
Model performance on testing set: 0.6164260950496623


#### Is `diabetes_target_test` exactly the same as the model prediction? Explain.

In [16]:
## No es exctamente igual. Porque como intenta predecir valores continuos, si diera resultado exactamente iguales significaria que tiene problema de overfitting

# Bonus Challenge 1 - Hypothesis Testing with `statsmodels`

After generating the linear regression model from the dataset, you probably wonder: then what? What is the statistical way to know if my model is reliable or not?

Good question. We'll discuss that using Scikit-Learn in Challenge 5. But for now, let's use a fool-proof way by using the ([Linear Regression class of StatsModels](https://www.statsmodels.org/dev/regression.html)) which can also conduct linear regression analysis plus much more such as calcuating the F-score of the linear model as well as the standard errors and t-scores for each coefficient. The F-score and t-scores will tell you whether you can trust your linear model.

To understand the statistical meaning of conducting hypothesis testing (e.g. F-test, t-test) for slopes, read [this webpage](https://onlinecourses.science.psu.edu/stat501/node/297/) at your leisure time. We'll give you a brief overview next.

* The F-test of your linear model is to verify whether at least one of your coefficients is significantly different from zero. Translating that into the *null hypothesis* and *alternative hypothesis*, that is:

    ```
    H0 : β1 = β2 = ... = β10 = 0
    HA : At least one βj ≠ 0 (for j = 1, 2, ..., 10)
    ```

* The t-tests on each coefficient is to check whether the confidence interval for the variable contains zero. If the confidence interval contains zero, it means the null hypothesis for that variable is not rejected. In other words, this particular vaiable is not contributing to your linear model and you can remove it from your formula.

Read the documentations of [StatsModels Linear Regression](https://www.statsmodels.org/dev/regression.html) as well as its [`OLS` class](https://www.statsmodels.org/dev/generated/statsmodels.regression.linear_model.OLS.html) which stands for *ordinary least squares*.

#### In the next cell, analyze `diabetes_data_train` and `diabetes_target_train` with the linear regression model of `statsmodels`. Print the fit summary.

Your output should look like:

![statsmodels regression](../statsmodels.png)

In [17]:
# Your code here:


#### Interpreting hypothesis testing results

Answer the following questions in the cell below:

1. What is the F-score of your linear model and is the null hypothesis rejected?

1. Does any of the t-tests of the coefficients produce a confidence interval containing zero? What are they?

1. How will you modify your linear reguression model according to the test results above?

In [18]:
# Your answers here:

# Challenge 3 - Peform Supervised Learning on a Pandas Dataframe

Now that we have dealt with data that has been formatted for scikit-learn, let's look at data that we will need to format ourselves.

In the next cell, load the `auto-mpg.csv` file included in this folder and assign it to a variable called `auto`.

In [19]:
auto = pd.read_csv('../auto-mpg.csv')

Look at the first 5 rows using the `head()` function:

In [20]:
auto.head()

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   
1  15.0          8         350.0        165.0    3693          11.5   
2  18.0          8         318.0        150.0    3436          11.0   
3  16.0          8         304.0        150.0    3433          12.0   
4  17.0          8         302.0        140.0    3449          10.5   

   model_year                       car_name  
0          70  \t"chevrolet chevelle malibu"  
1          70          \t"buick skylark 320"  
2          70         \t"plymouth satellite"  
3          70              \t"amc rebel sst"  
4          70                \t"ford torino"

Evaluate the data to ensure that all numeric columns are correctly detected as such by pandas. If a column is misclassified as object, coerce it to numeric.

In [21]:
auto.dtypes

mpg             float64
cylinders         int64
displacement    float64
horse_power     float64
weight            int64
acceleration    float64
model_year        int64
car_name         object
dtype: object

In [22]:
## no hay columnas con error de tipo de dato

What is the newest model year and the oldest model year?

In [23]:
auto['model_year'].min()
#modelo mas viejo

70

In [24]:
auto['model_year'].max()
#modelo mas nuevo

82

Check the dataset for missing values and remove all rows containing at least one missing value.

In [25]:
auto.isna().sum()

mpg             0
cylinders       0
displacement    0
horse_power     6
weight          0
acceleration    0
model_year      0
car_name        0
dtype: int64

In [26]:
auto.dropna(axis=0, inplace=True)
auto.isna().sum()

mpg             0
cylinders       0
displacement    0
horse_power     0
weight          0
acceleration    0
model_year      0
car_name        0
dtype: int64

Find the frequency table for the `cylinders` column using the `value_counts()` function. How many possible values of cylinders are there?

In [27]:
auto['cylinders'].value_counts()

4    199
8    103
6     83
3      4
5      3
Name: cylinders, dtype: int64

In [28]:
len(auto['cylinders'].value_counts())

5

In [29]:
## Existen 5 valores diferentes, es decir 5 tipos de cilindraje

We would like to generate a linear regression model that will predict mpg. To do this, first drop the `car_name` column since it does not contain any quantitative data. Next separate the dataframe to predictor and response variables. Separate those into test and training data with 80% of the data in the training set and the remainder in the test set. 

Assign the predictor and response training data to `X_train` and `y_train` respectively. Similarly, assign the predictor and response test data to `X_test` and `y_test`.

*Hint: To separate data for training and test, use the `train_test_split` method we used in previous labs.*

In [30]:
auto.drop('car_name', axis=1, inplace=True)

In [31]:
auto_model = LinReg()

In [32]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(auto.iloc[:, auto.columns != 'mpg'], auto['mpg'], test_size = 0.2)

In [33]:
auto_model.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [34]:
auto_model.predict(X_train)

array([31.21461813, 21.51826343, 19.2436121 , 24.30540906, 13.92374038,
       22.809079  , 18.02532604, 17.30272142, 24.71760424, 13.25576481,
       28.09847308, 24.54751062, 32.20735677, 26.25804923, 18.09612082,
       26.47690516, 21.68455262, 11.50850074, 22.972635  , 15.48751884,
       14.93277508, 26.41117133, 21.89799542, 24.06588353, 27.6765131 ,
       24.76536398, 22.3773503 , 34.48926637, 21.0080141 , 26.01113401,
       28.85809705, 30.64373513, 28.13997777, 24.08219748, 23.26835939,
       32.41509106, 26.53056142, 30.02237628, 11.95791378, 31.86315166,
       15.22978207, 22.94912471, 20.20203886, 15.49669293, 33.52278692,
       12.37561716, 24.65044849, 34.78917435, 21.83223999, 25.39213109,
       21.00785953, 33.20224545, 20.03980322, 12.97941592, 26.09097249,
       29.91227544,  9.91582073, 26.07736418, 11.21398017, 15.28749237,
       25.69166043, 32.78063604, 17.54981031, 12.02563078, 20.47138699,
       32.45872126, 27.21844357, 30.88781253, 29.2554248 , 22.39

In [35]:
mpg_preds = pd.DataFrame({'Real': y_test, 'Predicted': auto_model.predict(X_test)}).reset_index(drop=True)
mpg_preds.head(15)

Real  Predicted
0   12.0   8.115287
1   18.2  19.893207
2   17.5  21.029598
3   27.4  27.848836
4   19.0  22.027561
5   18.0  26.878581
6   16.0  17.359169
7   27.5  27.767960
8   21.0  22.483298
9   24.0  27.020563
10  20.0  23.878283
11  22.5  23.354297
12  27.9  27.731760
13  33.8  32.295875
14  20.2  24.368407

In [36]:
print(f"Model performance on trainng set: {auto_model.score(X_train, y_train)}")
print(f"Model performance on testing set: {auto_model.score(X_test, y_test)}")

Model performance on trainng set: 0.8118590235093095
Model performance on testing set: 0.7876040493542957


Now we will processed and peform linear regression on this data to predict the mpg for each vehicle. 

#### In the next cell, create an instance of the linear regression model and call it `auto_model`. Fit `auto_model` with your training data.

In [37]:
### Hecho arriba

# Challenge 4 - Evaluate the Model

In addition to evaluating your model with F-test and t-test, you can also use the *Coefficient of Determination* (a.k.a. *r squared score*). This method does not simply tell *yes* or *no* about the model fit but instead indicates how much variation can be explained by the model. Based on the r squared score, you can decide whether to improve your model in order to obtain a better fit.

You can learn about the r squared score [here](). Its formula is:

![R Squared](../r-squared.png)

...where:

* yi is an actual data point.
* ŷi is the corresponding data point on the estimated regression line.

By adding the squares of the difference between all yi-ŷi pairs, we have a measure called SSE (*error sum of squares*) which is an application of the r squared score to indicate the extent to which the estimated regression model is different from the actual data. And we attribute that difference to the random error that is unavoidable in the real world. Obviously, we want the SSE value to be as small as possible.

#### In the next cell, compute the predicted *y* based on `X_train` and call it `y_pred`. Then calcualte the r squared score between `y_pred` and `y_train` which indicates how well the estimated regression model fits the training data.

*Hint: r squared score can be calculated using `sklearn.metrics.r2_score` ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html)).*

In [38]:
from sklearn.metrics import r2_score

In [39]:
y_pred = auto_model.predict(X_train)

In [40]:
r2_score(y_pred, y_train)

0.7682590560151199

#### Our next step is to evaluate the model using the test data. 

We would like to ensure that our model is not overfitting the data. This means that our model was made to fit too closely to the training data by being overly complex. If a model is overfitted, it is not generalizable to data outside the training data. In that case, we need to reduce the complexity of the model by removing certain features (variables).

In the cell below, use the model to generate the predicted values for the test data and assign them to `y_test_pred`. Compute the r squared score of the predicted `y_test_pred` and the oberserved `y_test` data.

In [41]:
y_test_pred = auto_model.predict(X_test)

In [42]:
r2_score(y_test_pred, y_test)

0.7679143004554057

In [52]:
print(f"Model performance on trainng set: {auto_model.score(X_train, y_train)}")
print(f"Model performance on testing set: {auto_model.score(X_test, y_test)}")

Model performance on trainng set: 0.8118590235093095
Model performance on testing set: 0.7876040493542957


#### Explaining the results

The r squared scores of the training data and the test data are pretty close (0.8146 vs 0.7818). This means our model is not overfitted. However, there is still room to improve the model fit. Move on to the next challenge.

# Challenge 5 - Improve the Model Fit

While the most common way to improve the fit of a model is by using [regularization](https://datanice.github.io/machine-learning-101-what-is-regularization-interactive.html), there are other simpler ways to improve model fit. The first is to create a simpler model. The second is to increase the train sample size.

Let us start with the easier option and increase our train sample size to 90% of the data. Create a new test train split and name the new predictors and response variables `X_train09`, `X_test09`, `y_train09`, `y_test09`.

In [53]:
from sklearn.model_selection import train_test_split

X_train09, X_test09, y_train09, y_test09 = train_test_split(auto.iloc[:, auto.columns != 'mpg'], auto['mpg'], test_size = 0.1)

Initialize a new linear regression model. Name this model `auto_model09`. Fit the model to the new sample (training) data.

In [54]:
auto_model09 = LinReg()
auto_model09.fit(X_train09, y_train09)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

Compute the predicted values and r squared score for our new model and new sample data.

In [55]:
y_pred09 = auto_model09.predict(X_test09)

Compute the r squared score for the smaller test set. Is there an improvement in the test r squared?

In [56]:
print(f"Model performance on trainng set: {auto_model.score(X_train09, y_train09)}")
print(f"Model performance on testing set: {auto_model.score(X_test09, y_test09)}")

Model performance on trainng set: 0.8059545507511912
Model performance on testing set: 0.8243997314468802


In [57]:
## Si existe una minima mejora en el modelo

# Bonus Challenge 2 - Backward Elimination 

The main way to produce a simpler linear regression model is to reduce the number of variables used in the model. In scikit-learn, we can do this by using recursive feature elimination. You can read more about RFE [here](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFE.html).

In the next cell, we will import RFE

In [58]:
from sklearn.feature_selection import RFE
from sklearn.svm import SVR

Follow the documentation and initialize an RFE model using the `auto_model` linear regression model. Set `n_features_to_select=3`

In [60]:
rfe_model = RFE(auto_model, n_features_to_select=3)

Fit the model and print the ranking

In [63]:
selector = rfe_model.fit(auto.iloc[:, auto.columns != 'mpg'], auto['mpg'])

In [64]:
selector.ranking_

array([1, 2, 4, 3, 1, 1])

Feature importance is ranked from most important (1) to least important (4). Generate a model with the three most important features. The features correspond to variable names. For example, feature 1 is `cylinders` and feature 2 is `displacement`.

Perform a test-train split on this reduced column data and call the split data `X_train_reduced`, `X_test_reduced`, `y_test_reduced`, `y_train_reduced`. Use an 80% split.

Generate a new model called `auto_model_reduced` and fit this model. Then proceed to compute the r squared score for the model. Did this cause an improvement in the r squared score?

In [51]:
# Your code here: 


#### Conclusion

You may obtain the impression from this lab that without knowing statistical methods in depth, it is difficult to make major progress in machine learning. That is correct. If you are motivated to become a data scientist, statistics is the subject you must be proficient in and there is no shortcut. 

Completing these labs is not likely to make you a data scientist. But you will have a good sense about what are there in machine learning and what are good for you. In your future career, you can choose one of the three tracks:

* Data scientists who need to be proficient in statistical methods.

* Data engineers who need to be good at programming.

* Data integration specialists who are business or content experts but also understand data and programming. This cross-disciplinary track brings together data, technology, and business and will be in high demands in the next decade.